server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1597,105 +1624,30 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer, Entitlement enti
         Map<String, Integer> poolQuantityMap, CallerType caller)
         throws EntitlementRefusedException {
 
-        // Because there are several paths to this one place where entitlements
-        // are granted, we cannot be positive the caller obtained a lock on the
-        // pool when it was read. As such we're going to reload it with a lock
-        // before starting this process.
-        log.debug("Locking pools: {}", poolQuantityMap.keySet());
-        Collection<Pool> pools = poolCurator.lockAndLoadByIds(poolQuantityMap.keySet());
-
-        if (log.isDebugEnabled()) {
-            for (Pool pool : pools) {
-                log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
-            }
-        }
-
-        log.debug("Done locking pools");
+        BindChain chain = new BindChain();
+        BindContext context = bindContextFactory.create(consumer, poolQuantityMap);
+        chain.addOperation(rulesCheckOpFactory.create(caller));
+        chain.addOperation(handleEntitlementsOp);
+        chain.addOperation(postBindBonusPoolsOp);
+        chain.addOperation(handleCertificatesOp);
+        chain.addOperation(complianceOp);
 
```

Comment:
The BindChain is not really a 'chain'. It is basically a list of operations that get iterated over -- a command so to speak. IMO the chain should be responsible for dictating the pre/execute/post flow, not the pool manager itself. 

Taking this a little further, does the context really need to be exposed to the pool manager?

For consideration, we could execute the command and have it return the result, or throw the exception itself that will bubble up. All this would simply become the following in the pool manager, keeping everything encapsulated in the command/chain class.
```java
// Execute the command/chain with the provided args.
// A context related to this execution is created within the command/chain
// and never gets exposed outside of a single execution.
List<Entitlement> ents = bindcommand.execute(consumer, poolQuantityMap);
```